packages/devtools_app/lib/src/screens/memory/panes/leaks/diagnostics/model.dart

```diff
@@ -0,0 +1,162 @@
+import 'package:vm_service/vm_service.dart';
+
+import '../instrumentation/model.dart';
+
+/// Result of analysis of [notGCed] memory leaks.
+class NotGCedAnalyzed {
+  NotGCedAnalyzed(this.byCulprits, this.withoutPath, this.total);
+
+  /// Not GCed objects withretaining path to the root, by culprits.
+  final Map<LeakReport, List<LeakReport>> byCulprits;
+
+  /// Not GCed objects without retaining path to the root.
+  final List<LeakReport> withoutPath;
+
+  /// Total number of leaks.
+  final int total;
+}
+
+/// Input for analyses of notGCed leaks.
+class NotGCedAnalyzerTask {
+  NotGCedAnalyzerTask({
+    required this.heap,
+    required this.reports,
+  });
+
+  factory NotGCedAnalyzerTask.fromJson(Map<String, dynamic> json) =>
+      NotGCedAnalyzerTask(
+        reports: (json['reports'] as List<dynamic>)
+            .cast<Map<String, dynamic>>()
+            .map((e) => LeakReport.fromJson(e))
+            .toList(),
+        heap: AdaptedHeap.fromJson(json['heap']),
+      );
+
+  NotGCedAnalyzerTask.fromSnapshot(HeapSnapshotGraph graph, this.reports)
+      : heap = AdaptedHeap.fromHeapSnapshot(graph);
+
+  final AdaptedHeap heap;
+  final List<LeakReport> reports;
+
+  Map<String, dynamic> toJson() => {
+        'reports': reports.map((e) => e.toJson()).toList(),
+        'heap': heap.toJson(),
+      };
+}
+
+/// Contains information from [HeapSnapshotGraph], necessary to analyze
+/// memory leaks, plus serialization.
+class AdaptedHeap {
+  AdaptedHeap(this.objects);
+
+  factory AdaptedHeap.fromJson(Map<String, dynamic> json) => AdaptedHeap(
+        (json['objects'] as List<dynamic>)
+            .map((e) => AdaptedHeapObject.fromJson(e))
+            .toList(),
+      );
+
+  factory AdaptedHeap.fromHeapSnapshot(HeapSnapshotGraph graph) => AdaptedHeap(
+        graph.objects
+            .map((e) => AdaptedHeapObject.fromHeapSnapshotObject(e))
+            .toList(),
+      );
+
+  static const rootIndex = 1;
+  final List<AdaptedHeapObject> objects;
+  bool isSpanningTreeBuilt = false;
+
+  late final Map<IdentityHashCode, int> _byCode = Map.fromIterable(
+    Iterable.generate(objects.length),
+    key: (i) => objects[i].code,
+    value: (i) => i,
+  );
+
+  Map<String, dynamic> toJson() => {
+        'objects': objects.map((e) => e.toJson()).toList(),
+      };
+
+  HeapPath? _retainingPath(IdentityHashCode code) {
+    assert(isSpanningTreeBuilt);
+    var i = _byCode[code]!;
+    if (objects[i].retainer == null) return null;
+
+    final result = <int>[];
```

Comment:
we could either use a queue to always insert at the front, or we could still use a list, but `add` to the list instead of `insert`ing, and then `return result.reversed` at the end.